File: leetcode/src/AddTwoNumbers.java
import java.util.ArrayList;
import java.util.List;

public class AddTwoNumbers {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode result = null;
        List<Integer> resultList = new ArrayList<Integer>();
        ListNode current1 = l1;
        ListNode current2 = l2;
        int add = 0;
        while ((current1 != null) || (current2 != null)) {
            int item1 = current1 == null ? 0 : current1.val;
            int item2 = current2 == null ? 0 : current2.val;
            int sum = item1 + item2 + add;
            resultList.add(sum % 10);
            add = sum / 10;
            if (current1 != null) current1 = current1.next;
            if (current2 != null) current2 = current2.next;
        }
        if (add > 0) {
            resultList.add(add);
        }
        for (int i = resultList.size() - 1; i >= 0; i--) {
            if (result == null) {
                result = new ListNode(resultList.get(i));
            } else {
                result = new ListNode(resultList.get(i), result);
            }
        }
        return result;
    }
}

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

File: leetcode/src/python/add_two_numbers.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class AddTwoNumbers:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        root = None
        current = None
        leftover = 0
        while l1 is not None or l2 is not None:
            sum = leftover
            sum += l1.val if l1 is not None else 0
            sum += l2.val if l2 is not None else 0
            if sum >= 10:
                leftover = 1
                sum = sum - 10
            else:
                leftover = 0
            if root is None:
                root = ListNode(sum)
                current = root
            else:
                current.next = ListNode(sum)
                current = current.next
            l1 = l1.next if l1 is not None else l1
            l2 = l2.next if l2 is not None else l2
        if leftover != 0:
            current.next = ListNode(leftover)
        return root